aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,467 @@
+              "dim must be non-negative and less than input dimensions");
+  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "softmax", [&] {
+      // assuming that the type of input._indices() entries is int64_t
+      cpu_sparse_coo_softmax<scalar_t, int64_t, false>(output, input, dim_);
+  });
+  return output;
+}
+
+Tensor log_softmax_sparse_cpu(const Tensor& input_, const int64_t dim_, const bool half_to_float) {
+  AT_ASSERT(input_.is_sparse());
+  AT_ASSERTM(!half_to_float, "log_softmax with half to float conversion is not supported on CPU");
+  auto input = input_.coalesce();
+  Tensor output = at::native::empty_like(input);
+  if (input.numel() == 0) {
+    return output;
+  }
+  TORCH_CHECK(dim_ >= 0 && dim_ < input.dim(),
+              "dim must be non-negative and less than input dimensions");
+  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "log_softmax", [&] {
+      // assuming that the type of input._indices() entries is int64_t
+      cpu_sparse_coo_softmax<scalar_t, int64_t, true>(output, input, dim_);
+  });
+  return output;
+}
+
+Tensor softmax_backward_sparse_cpu(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& input_) {
+  TensorArg grad_arg{grad_, "grad", 1}, output_arg{output_, "output", 2};
+  checkSameSize("softmax_backward", grad_arg, output_arg);
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
+  auto output = output_.coalesce();
+
+  Tensor grad_input = at::native::empty_like(output);
+  if (output.numel() == 0) {
+    return grad_input;
+  }
+  TORCH_CHECK(
+      dim >= 0 && dim < grad.dim(),
+      "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(
+              grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
+      // assuming that the type of input._indices() entries is int64_t
+      cpu_sparse_coo_softmax_backward<scalar_t, int64_t, false>(grad_input, grad, output, dim);
+  });
+  return grad_input;
+}
+
+Tensor log_softmax_backward_sparse_cpu(
+    const Tensor& grad_,
+    const Tensor& output_,
+    int64_t dim_,
+    const Tensor& input_) {
+  TensorArg grad_arg{grad_, "grad", 1}, output_arg{output_, "output", 2};
+  checkSameSize("log_softmax_backward", grad_arg, output_arg);
+
+  int64_t dim = maybe_wrap_dim(dim_, grad_.dim());
+
+  auto grad = grad_.coalesce();
+  auto output = output_.coalesce();
+
+  Tensor grad_input = at::native::empty_like(output);
+  if (output.numel() == 0) {
+    return grad_input;
+  }
+  TORCH_CHECK(
+      dim >= 0 && dim < grad.dim(),
+      "dim must be non-negative and less than input dimensions");
+  TORCH_CHECK(
+              grad.sparse_dim() == output.sparse_dim(),
+      "grad and output sparse dimensions must be equal");
+  AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "softmax_backward", [&] {
```

Comment:
I guess the `dtype` dispatching would take place here.

Reply:
yes